javascript/modules/client/src/internal/encoding/client.ts

```diff
@@ -4,13 +4,131 @@ import { BigNumber } from "@ethersproject/bignumber";
 import { AddressZero } from "@ethersproject/constants";
 import { AVAILABLE_CLIENT_FUNCTION_SIGNATURES } from "../constants/encoding";
 import { FunctionFragment, Interface, Result } from "@ethersproject/abi";
-import { IWithdrawParams } from "../interfaces/client";
+import {
+  DaoPermissions,
+  IFreezeParams,
+  IPermissionParams,
+  IWithdrawParams,
+} from "../interfaces/client";
 
 export function getFunctionFragment(data: Uint8Array): FunctionFragment {
   const hexBytes = bytesToHex(data, true);
   const inter = new Interface(AVAILABLE_CLIENT_FUNCTION_SIGNATURES);
   return inter.getFunction(hexBytes.substring(0, 10));
 }
+export function encodeFreezeAction(
+  params: IFreezeParams,
+): Uint8Array {
+  const daoInterface = DAO__factory.createInterface();
+  const args = unwrapFreezeParams(params);
+  // get hex bytes
+  const hexBytes = daoInterface.encodeFunctionData("freeze", args);
+  return hexToBytes(strip0x(hexBytes));
+}
+
+export function decodeFreezeActionData(data: Uint8Array): IFreezeParams {
+  const daoInterface = DAO__factory.createInterface();
+  const hexBytes = bytesToHex(data, true);
+  const receivedFunction = daoInterface.getFunction(
+    hexBytes.substring(0, 10) as any,
+  );
+  const expectedFunction = daoInterface.getFunction("freeze");
+  if (receivedFunction.name !== expectedFunction.name) {
+    throw new Error(
+      "The received action is different from the expected one",
+    );
+  }
+  const result = daoInterface.decodeFunctionData("freeze", data);
+  return wrapFreezeParams(result);
+}
+
+function unwrapFreezeParams(
+  params: IFreezeParams,
+): [string, string] {
+  return [
+    params.where,
+    params.permission,
+  ];
+}
+function wrapFreezeParams(
+  result: Result,
+): IFreezeParams {
+  return {
+    where: result[0],
+    permission: result[1] as DaoPermissions,
+  };
+}
+
+export function encodeGrantActionData(
+  params: IPermissionParams,
+): Uint8Array {
+  const daoInterface = DAO__factory.createInterface();
+  const args = unwrapPermissionParams(params);
+  // get hex bytes
+  const hexBytes = daoInterface.encodeFunctionData("grant", args);
+  return hexToBytes(strip0x(hexBytes));
+}
+
+export function decodeGrantActionData(data: Uint8Array): IPermissionParams {
+  const daoInterface = DAO__factory.createInterface();
+  const hexBytes = bytesToHex(data, true);
+  const receivedFunction = daoInterface.getFunction(
+    hexBytes.substring(0, 10) as any,
+  );
+  const expectedFunction = daoInterface.getFunction("grant");
+  if (receivedFunction.name !== expectedFunction.name) {
+    throw new Error(
+      "The received action is different from the expected one",
+    );
+  }
+  const result = daoInterface.decodeFunctionData("grant", data);
+  return wrapPermissionParams(result);
```

Comment:
Shouldn't we do a loop here, to find which of the knows PERMISSION ID's is been requested?